### PR TITLE
esp8266: serial swap & softserial + debug on separate tcp server while printing

### PR DESCRIPTION
**Pull request by d-a-v**
*edit2:* tcp debug server while printing, on default port 8889 (when debug on tcp is enabled)

option `SERIAL_SWAP` to enable in `config.h`:
- **no more serial garbage** (at boot or on failure) by using Hardware serial on GPIO13/15 for printer
- allow debugging on serial via USB while printing (using software serial output)

also,
- when using esp8266 arduino core 2.4.2 and git master, let default AP IP address (then it works)

and,
- fixed compilation warnings
- convert tabs to spaces
- remove trailing spaces

Note:
This is not tested on my printer yet (emotion-tech / microdelta-rework / repetier),
but an USB/serial converter on GPIO13/15 shows only G-codes while my arduino
serial console shows all debugs messages (and serial garbage at boot time).

*edit*:
Maybe for reviewing it is better to check commits one by one.
Sorry for putting all of this in one PR (I can propose multiple PR is desired)

### CODE REVIEW

**Comment by luc-github** [comment]
Thank you - Merci du coup de patte ^_^
Yes  long time I did not applied astyle to clean code
I will review the PR soon as I am travelling and I only bring one ESP01 with me.
Actually current boot garbage is not an issue as printer power up  is always slower than ESP, crash are an issue of course, but should not happen thanks to great core maintainers ;)

File: esp3d/wificonf.cpp [hunk 1]
```diff
@@ -185,6 +187,16 @@ bool WIFI_CONFIG::Setup(bool force_ap)
         LOG("SSID ")
         LOG(sbuf)
         LOG("\r\n")
+
+#ifndef LWIP_VERSION_MAJOR
+#error
+#endif
+#if defined(ARDUINO_ARCH_ESP32) || LWIP_VERSION_MAJOR == 1
```
Comment (luc-github) [inline esp3d/wificonf.cpp]: I was not aware setting IP on AP is not working anymore with LWIP 2 
Also for me it works on ARDUINO_ARCH_ESP32

Comment (d-a-v) [reply esp3d/wificonf.cpp]: Well, this was not a known bug (at least to me) until now.
Once fixed, and if this is merged as is, I will revisit here to update that.

**Comment by d-a-v** [comment]
I'm sorry to have included this astyle job.
I just started the tab change around the code I was reading and,
once started, I just could not decently deliver an half finished work.

I already understood that the serial garbage was not an issue anymore for you.
Before I discovered your project,  I wanted to make a clean tcp<->serial passthrough for my printer.
I have been wondering how to avoid this issue, and `Serial.swap()` came to me as an obvious answer (not so obvious for a long time though). As I find your project very successful I thought that could be a nice optional addition.

> but should not happen thanks to great core maintainers ;)

Not so great, I came after core-2.3.0 and you are still advising to use core-2.2.0 ;-)
Is core-2.3.0 unstable ? I know about 2.4.0 and 2.4.1 instabilities.
I can't tell if ESP3D is stable enough with core-2.4.2 (or git version) because I'm a too young ESP3D user, but you tell me. I will be happy to try and fix anything that prevent its use in place of previous versions.

**Comment by luc-github** [comment]
No no it is ok - thank you

I just have a concern  for initial swap() because the project also can use pin 13 for SPI SD card  (code is not yet public) but as it is an option it should not be an issue.

it is not that 2.3 is unstable but I did not review all feature with it and I did not have enough step back then 2.4 popup with several problems. 
And I have discovered some chinese company still use esp core 1.0 + ESP3D 0.5 to sell their devices...

So far 2.2 is working without crash for ESP3D so the release (do not fix what is not broken).

The git advise to use the git and latest ESP core to get feedback and so to do a release, but their are too many changes in small amouth of time, I need to be sure it is stable because in same time ESP3D has new versions and feature, testing all take time, but hope to do one release soon 2.0 is still in Devt and use Git ESP core  and 2.1 is now on GRBL/Marlin. I have hard time to follow myself :P 

You feedback will be valuable for sure

**Comment by d-a-v** [comment]
> And I have discovered some chinese company still use esp core 1.0 + ESP3D 0.5 to sell their devices...

congrats for that !
and for the 300 :star: 

> I just have a concern for initial swap() because the project also can use pin 13 for SPI SD card (code is not yet public) but as it is an option it should not be an issue.

Ah true!
Then serial swapping will be disabled for SD card users.
This PR is not a showstopper anyway, only useful for the peace of mind of users who know what's going on on the serial link at boot time ;)

I have to revisit the AP address issue,
Are you ok for leaving this opened for a little while?

**Comment by luc-github** [comment]
Thank you ^_^

yes sure - I am travelling right now and no real time to sit and review carefully so no worry

**Comment by luc-github** [comment]
Looks like compilation failed for ESP32 :
`error: 'class WiFiClient' has no member named 'availableForWrite'`
and as discussed PR is for 1.0 when devt is 2.0 and now move to 2.1 - so I will do the changes by hand
Until then I left this PR open

**Comment by d-a-v** [comment]
Closing this PR to keep a clean repository.
I will come back to this at some point.

**Comment by luc-github** [comment]
Actually I have already integrated almost all suggestion in 3.0 - just - software serial is not yet on it

**Comment by luc-github** [comment]
https://github.com/all-contributors please add @d-a-v for code and idea improvement

**Comment by luc-github** [comment]
@all-contributors please add @d-a-v for code and idea improvement

**Comment by luc-github** [comment]
@all-contributors please add @d-a-v for code and idea improvement

**Pull request closed without merging**